### ISSUE TITLE
Full automatic app update option in setting Gui

### ISSUE THREAD

**Issue by bd8392** [nextcloud/server#15416]
Please implement an option that all apps (or just activatet) update themselves at a certain time...this would be very useful! Thanks :-)

**Comment by bpcurse** [comment]
Seems to be a duplicate of #15360.
Or is it an enhancement to it as a time-trigger?

**Comment by bd8392** [comment]
Would say enhancement...cause update time is very important in productive environments...

**Comment by wiswedel** [comment]
>Would say enhancement...cause update time is very important in productive environments...

Agreed. An "update all" button might be a compromise (like in the iOS App Store). 
What do you think?

**Comment by bd8392** [comment]
Would be good, but not the solution for me....I use nexrtcloud as docker with watchtower....so it gets updated automatically (except the apps). The best solution would be the followwing: Nextcloud should watch the apps....if there is an update, nextcloud should make it at a given time. Also when an update of nextcloud itself is performed, it should detect which apps got deactivated because of the update....then nextcloud should update them and then reactivate them (during nextcloud update). This would fully automate every nextcloudupdate (which would be really great!)

**Comment by wiswedel** [comment]
You could write a script to update apps via occ and let it run as cronjob. How about that?

**Comment by wiswedel** [comment]
https://github.com/nextcloud/server/pull/11053

**Comment by bd8392** [comment]
Good idea but still not the optimal solution because it is not very end-user friendly...the other problem is, if nextcloud gets updated, certain apps get deactivated...the Cron-job wouldn't reactivate them...but do you have such a script for testing?

**Comment by wiswedel** [comment]
> Good idea but still not the optimal solution because it is not very end-user friendly

Full-automation doesn't have to be user-friendly, does it?

**Comment by bd8392** [comment]
It would be a lot easier...also your solution does not solve the app problem...can't you consider it again?

**Comment by wiswedel** [comment]
As this sounds like a nice feature, the requests for this are quite low. Currently there are no plans to implement such a feature. Thus I will close this ticket for now. This does not mean we don't want this feature, but it is simply not on our roadmap for the near future. If somebody wants to implement this feature nevertheless we are happy to assist and help out.

If you wish to have this feature implemented by the Nextcloud GmbH there is the option for consulting work on top of your Nextcloud Enterprise subscription to get your features implemented.

**Comment by bd8392** [comment]
Okay :-) thanks for your response :-)

**Comment by georgehrke** [comment]
@MorrisJobke @rullzer Isn't this something we eventually wanted to have on major upgrades?

**Comment by MorrisJobke** [comment]
> @MorrisJobke @rullzer Isn't this something we eventually wanted to have on major upgrades?

We check for updates if they are needed. cc @nickvergessen

**Comment by georgehrke** [comment]
> We check for updates if they are needed.

We had this problem with the calendar app and Nextcloud 16. Calendar 1.6.x was only compatible with Nextcloud 14 and 15. Calendar 1.7.x is only compatible with Nextcloud 16.

Instead of automatically updating to Calendar 1.7.x, the calendar app was disabled, labelled as incompatible and admins had to update it first, before they were able to enable it again.

Causing quite some bug reports in the calendar app, because users claimed there is no compatible version of the Calendar with Nextcloud 16.

**Comment by nickvergessen** [comment]
That should™ work, the update screen shows the warning if no update is available. Also I remember that there was code in 16 to check from the appstore and reenable apps after updating them from there.

**Comment by the-sane** [comment]
@nickvergessen @MorrisJobke What you describe does not appear to be working as intended.  It did not happen on my instance, @georgehrke just described it failing for other users with the calendar app, and the snap v16 release continues to be held up by this problem.  @kyrofa, the snap maintainer, is currently in the process of writing his own post-update process to handle this situation.  Please see the issue referenced directly above.  Perhaps someone wants to pop over there and let him know if it really is just a bug?

**Comment by kyrofa** [comment]
Thanks for the ping @the-sane, this is interesting. @nickvergessen it's pretty easy to verify: install v15, install the calendar, update to v16, run `occ upgrade`, watch calendar get disabled (and not updated/re-enabled). I agree, it probably makes sense for Nextcloud to try and update incompatible apps on its own instead of what it currently seems to be doing. It would certainly save us some work, too :) .

**Comment by georgehrke** [comment]
There is indeed code for it: 
https://github.com/nextcloud/server/blob/master/lib/private/Updater.php#L258L259

Maybe `\OC_App::$autoDisabledApps` is not populated correctly.

Treating this as a bug now rather than an enhancement. Will check tomorrow.

**Comment by georgehrke** [comment]
Okay, so my assumption was correct:
- `\OC_App::$autoDisabledApps` stays empty
- instead, the app is disabled here:
https://github.com/nextcloud/server/blob/7e7146db7fdefa7039ad5dd9653500589a1d0c51/lib/private/Updater.php#L407
- as a consequence, `\OC_App::loadApp`, the only place where `\OC_App::$autoDisabledApps` is populated, is never called.

**Comment by georgehrke** [comment]
Possible fix in https://github.com/nextcloud/server/pull/16507
